File: player2.py
import time
from enum import Enum
from datetime import datetime, timedelta

class PlayerStatus(Enum):
    NORMAL = "normal"
    STOPPED = "stopped"
    SLOW_MOTION = "in slow motion"
    INVISIBLE = "invisible"
    SPEED_BOOST = "speed++"
    BLOCKED = "blocked by an obstacle"

class PowerType(Enum):
    BONUS = "bonus"
    MALUS = "malus"

class Power:
    def __init__(self, name, power_type, casting_time, cooldown):
        self.name = name
        self.type = power_type
        self.casting_time = casting_time  # in seconds
        self.cooldown = cooldown          # in seconds
        self.last_used_time = datetime.min  # Last time the power was used

    def activate(self, target):
        current_time = datetime.now()
        time_since_last_use = (current_time - self.last_used_time).total_seconds()

        if time_since_last_use < self.cooldown:
            print(f"{self.name} is still in cooldown. Wait {self.cooldown - time_since_last_use:.2f} seconds.")
            return

        print(f"Preparing {self.name}... (casting time: {self.casting_time} seconds)")
        time.sleep(self.casting_time)  # Simulating casting time

        # Power is now activated
        print(f"{self.name} activated on {target.name}!")
        self.last_used_time = datetime.now()  # Reset cooldown timer

        if self.type == PowerType.BONUS:
            target.apply_bonus(self)
        elif self.type == PowerType.MALUS:
            target.apply_malus(self)

class Player:
    def __init__(self, name):
        self.name = name
        self.status = PlayerStatus.NORMAL

    def apply_bonus(self, power):
        print(f"{self.name} receives a bonus from {power.name}!")
        # Apply the effect of the bonus here (e.g., change player status)

    def apply_malus(self, power):
        print(f"{self.name} is affected by a malus from {power.name}!")
        # Apply the effect of the malus here (e.g., slow down the player)

class Evilman:
    def __init__(self, name="Evilman"):
        self.name = name
        # Powers :
        self.slowmotion_power = Power("Slow Motion Trap", PowerType.MALUS, 3, 8)  # 3 seconds casting, 8 seconds cooldown
        self.block_power = Power("Block Path", PowerType.MALUS, 2, 10)       # 2 seconds casting, 10 seconds cooldown
        self.stop_power = Power("Stop", PowerType.MALUS, 5, 10)              # 5 seconds casting, 10 seconds cooldown 

    def use_slowmotion_power(self, player):
        print(f"{self.name} tries to use malus power: {self.malus_power.name} on {player.name}")
        self.slowmotion_power.activate(player)

    def use_block_power(self, player):
        print(f"{self.name} tries to block {player.name}'s path with {self.block_power.name}")
        self.block_power.activate(player)
    
    def use_stop_power(self, player):
        print(f"{self.name} tries to stop {player.name} with {self.stop_power.name}")
        self.stop_power.activate(player)

class Protector:
    def __init__(self, name="Protector"):
        self.name = name
        # Powers :
        self.speed_power = Power("Speed Boost", PowerType.BONUS, 2, 5)                  # 2 seconds casting, 5 seconds cooldown
        self.invisibleshield_power = Power("Invisible Shield", PowerType.BONUS, 1, 7)   # 1 second casting, 7 seconds cooldown
        self.stop_evilman_power = Power("Stop Evilman", PowerType.BONUS, 5, 10)         # 5 second casting, 10 seconds cooldown

    def use_speed_power(self, player):
        print(f"{self.name} tries to use bonus power: {self.speed_power.name} on {player.name}")
        self.speed_power.activate(player)

    def use_invisibleshield_power(self, player):
        print(f"{self.name} tries to protect {player.name} with {self.invisibleshield_power.name}")
        self.invisibleshield_power.activate(player)
    
    def use_stop_evilman_power(self, player):
        print(f"{self.name} tries to protect {player.name} with {self.stop_evilman_power.name}")
        self.stop_evilman_power.activate(player)

# Main code to simulate interaction
def main():
    # Create actors
    player = Player("Alex")
    evilman = Evilman()
    protector = Protector()

    # Evilman uses a malus power on the player
    evilman.use_slowmotion_power(player)  # Slow motion trap, 3 seconds casting, 8 seconds cooldown
    time.sleep(4)                    # Simulating time passage
    evilman.use_slowmotion_power(player)  # Still in cooldown

    # Protector uses a bonus power on the player
    protector.use_speed_power(player)  # Speed Boost, 2 seconds casting, 5 seconds cooldown
    time.sleep(6)                      # Simulating time passage
    protector.use_speed_power(player)  # Can be used again after cooldown

    # Evilman tries to block the player
    evilman.use_block_power(player)     # Block path, 2 seconds casting, 10 seconds cooldown

    # Protector protects the player with a shield
    protector.use_invisibleshield_power(player)  # Invisible shield, 1 second casting, 7 seconds cooldown

    # Evilman tries to stop the player
    evilman.use_stop_power(player)     # Player stopped, 5 seconds casting, 10 seconds cooldown

    # Protector helps the player stopping Evilman
    protector.use_stop_evilman_power(player)  # Stop Evilman, 5 second casting, 10 seconds cooldown



if __name__ == "__main__":
    main()

File: player2.cs
using System;
using System.Threading.Tasks;  // For Task.Delay to simulate time.sleep()

public enum PlayerStatus
{
    NORMAL,
    STOPPED,
    SLOW_MOTION,
    INVISIBLE,
    SPEED_BOOST,
    BLOCKED
}

public enum PowerType
{
    BONUS,
    MALUS
}

public class Power
{
    public string Name { get; }
    public PowerType Type { get; }
    public int CastingTime { get; }    // Casting time in milliseconds
    public int Cooldown { get; }       // Cooldown in milliseconds
    private DateTime lastUsedTime;

    public Power(string name, PowerType type, int castingTime, int cooldown)
    {
        Name = name;
        Type = type;
        CastingTime = castingTime;
        Cooldown = cooldown;
        lastUsedTime = DateTime.MinValue;  // Initialized to a time in the past
    }

    public async Task Activate(Player target)
    {
        var currentTime = DateTime.Now;
        var timeSinceLastUse = (currentTime - lastUsedTime).TotalMilliseconds;

        if (timeSinceLastUse < Cooldown)
        {
            Console.WriteLine($"{Name} is still in cooldown. Wait {(Cooldown - timeSinceLastUse) / 1000.0:F2} seconds.");
            return;
        }

        Console.WriteLine($"Preparing {Name}... (casting time: {CastingTime / 1000.0:F2} seconds)");
        await Task.Delay(CastingTime);  // Simulating casting time

        // Power is now activated
        Console.WriteLine($"{Name} activated on {target.Name}!");
        lastUsedTime = DateTime.Now;  // Reset cooldown timer

        if (Type == PowerType.BONUS)
        {
            target.ApplyBonus(this);
        }
        else if (Type == PowerType.MALUS)
        {
            target.ApplyMalus(this);
        }
    }
}

public class Player
{
    public string Name { get; }
    public PlayerStatus Status { get; set; }

    public Player(string name)
    {
        Name = name;
        Status = PlayerStatus.NORMAL;  // Player starts in a normal state
    }

    public void ApplyBonus(Power power)
    {
        Console.WriteLine($"{Name} receives a bonus : {power.Name}!");
        // Apply the effect of the bonus here, such as changing the player's status
    }

    public void ApplyMalus(Power power)
    {
        Console.WriteLine($"{Name} is affected by a malus : {power.Name}!");
        // Apply the effect of the malus here, such as slowing down the player
    }
}

public class Evilman
{
    public string Name { get; }
    public Power MalusPower { get; }
    public Power BlockPower { get; }

    public Evilman(string name = "Evilman")
    {
        Name = name;
        SlowMotionPower = new Power("Slow Motion Trap", PowerType.MALUS, 3000, 8000);  // 3 seconds casting, 8 seconds cooldown
        BlockPower = new Power("Block Path", PowerType.MALUS, 2000, 10000);        // 2 seconds casting, 10 seconds cooldown
    }

    public async Task UseSlowMotionPower(Player player)
    {
        Console.WriteLine($"{Name} tries to use Slow Motion power: {SlowMotionPower.Name} on {player.Name}");
        await SlowMotionPower.Activate(player);
    }

    public async Task UseBlockPower(Player player)
    {
        Console.WriteLine($"{Name} tries to block {player.Name}'s path with {BlockPower.Name}");
        await BlockPower.Activate(player);
    }
}

public class Protector
{
    public string Name { get; }
    public Power BonusPower { get; }
    public Power ShieldPower { get; }

    public Protector(string name = "Protector")
    {
        Name = name;
        SpeedPower = new Power("Speed Boost", PowerType.BONUS, 2000, 5000);      // 2 seconds casting, 5 seconds cooldown
        InvisibleShieldPower = new Power("Invisible Shield", PowerType.BONUS, 1000, 7000); // 1 second casting, 7 seconds cooldown
    }

    public async Task UseSpeedPower(Player player)
    {
        Console.WriteLine($"{Name} tries to use Speed Boost power: {SpeedPower.Name} on {player.Name}");
        await SpeedPower.Activate(player);
    }

    public async Task UseInvisibleShieldPower(Player player)
    {
        Console.WriteLine($"{Name} tries to protect {player.Name} with {InvisibleShieldPower.Name}");
        await InvisibleShieldPower.Activate(player);
    }
}

public class Program
{
    public static async Task Main(string[] args)
    {
        // Creating actors
        Player player = new Player("Alex");
        Evilman evilman = new Evilman();
        Protector protector = new Protector();

        // Evilman uses a malus power on the player
        await evilman.UseSlowMotionPower(player);  // Slow motion trap, 3 seconds casting, 8 seconds cooldown
        await Task.Delay(4000);               // Simulating time passage = 4 seconds
        await evilman.UseSlowMotionPower(player);  // Still in cooldown

        // Protector uses a bonus power on the player
        await protector.UseSpeedPower(player);  // Speed Boost, 2 seconds casting, 5 seconds cooldown
        await Task.Delay(6000);                 // Simulating time passage = 6 seconds
        await protector.UseSpeedPower(player);  // Can be used again after cooldown

        // Evilman tries to block the player
        await evilman.UseBlockPower(player);    // Block path, 2 seconds casting, 10 seconds cooldown

        // Protector protects the player with a shield
        await protector.UseInvisibleShieldPower(player); // Invisible shield, 1 second casting, 7 seconds cooldown
    }
}